File: Graph/BFS.py
# BFS
# 임의의 노드에서 시작하여 인접한 노드를 모두 확인한 후 다음 depth를 탐색
# 큐 자료구조 활용

graph = [[2, 1], [3, 0], [3, 0], [9, 8, 2, 1], [5], [7, 6, 4], [7, 5], [6, 5], [3], [3]]
N = len(graph)
visited = [False] * N


def bfs(start):
    queue = []
    queue.append(start)
    visited[start] = True
    while len(queue) != 0:
        v = queue.pop(0)
        print(v, '', end='')
        for i in graph[v]:
            if not visited[i]:
                visited[i] = True
                queue.append(i)

print('BFS 방문 순서: ', end='')
for i in range(N):
    if not visited[i]:
        bfs(i)

File: data_structure_js/graph/bfs.js
// ref. https://chamdom.blog/dfs-using-js/

function bfs(graph, start) {
  const queue = [];
  const visited = [];
  const result = [];

  queue.push(start);
  visited[start] = true;
  result.push(start);

  while (queue.length) {
    const node = queue.shift();

    for (const el of graph[node]) {
      if (visited[el]) continue;

      queue.push(el);
      visited[el] = true; // 방문체크는 큐에서 꺼낼 때가 아니라 넣을 때 하자
      result.push(el);
    }
  }

  return result;
}

// 인접 리스트 방식
// 정점이 여러개일때 낮은 정점부터 방문해야 할 경우 오름차순 정렬한다.
const graph = [[1, 2, 4], [0, 5], [0, 5], [4], [0, 3], [1, 2]];

const bfsResult = bfs(graph, 0);
console.log(bfsResult.join(' ')); // 0 1 2 4 5 3
